### ISSUE TITLE
TextMatcherEnvelope should call `asString()` only once at all

### ISSUE THREAD

**Issue by victornoel** [llorllale/cactoos-matchers#132]
All the `Matcher<Text>` that relies on `TextMatcherEnvelope` suffers from having `asString()` being called multiple times.

In some case it can be a problem, for example when a `Text` is based on an `InputStream` that can be read only once. See for example https://github.com/yegor256/cactoos/pull/1145#discussion_r296447726

I believe `TextMatcherEnvelope` should wraps `text` into some kind of `Sticky` decorator to avoid any problem.

**Comment by 0crat** [comment]
@llorllale[/z](https://www.0crat.com/u/llorllale) please, pay attention to this issue

**Comment by llorllale** [comment]
@victornoel funny - I was just reading [this](https://www.pragmaticobjects.com/chapters/008_things-i-would-never-place-in-decorators.html#caching-and-memoization) today. I'm closing this issue and deferring this conversation for later. I'll accept your PR in yegor256/cactoos#1145 as it

**Comment by 0crat** [comment]
Job `gh:llorllale/cactoos-matchers#132` is not assigned, can't get performer

**Comment by 0crat** [comment]
This job is not in scope

**Comment by victornoel** [comment]
@llorllale I've read it too, I don't agree with the content personally. That being said, I don't think this issue is about caching or not caching, it is just about having a consistent behaviour of `Text` matchers, whatever the mean used to solve the problem.

**Comment by llorllale** [comment]
@victornoel what do you propose? Wrap it in `Sticky` in `TextMatcherEnvelope#matchesSafely()`?

**Comment by victornoel** [comment]
@llorllale I would have done something like that yes I think

**Comment by llorllale** [comment]
@0crat in

**Comment by 0crat** [comment]
@llorllale[/z](https://www.0crat.com/u/llorllale) please, pay attention to this issue

**Comment by 0crat** [comment]
> @0crat in ([here](https://github.com/llorllale/cactoos-matchers/issues/132#issuecomment-508303973))

@llorllale Job [#132](https://github.com/llorllale/cactoos-matchers/issues/132) is now in scope, role is `DEV`

**Comment by 0crat** [comment]
Bug was reported, see [§29](http://www.zerocracy.com/policy.html#29): +15 point(s) just awarded to @victornoel[/z](https://www.0crat.com/u/victornoel)

**Comment by 0crat** [comment]
The job [#132](https://github.com/llorllale/cactoos-matchers/issues/132) assigned to @umbrah[/z](https://www.0crat.com/u/umbrah), here is [why](https://www.0crat.com/footprint/C63314D6Z/46044a7b-aded-4372-b37a-5bc39cd32932); the budget is 30 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§8](http://www.zerocracy.com/policy.html#8) and [§9](http://www.zerocracy.com/policy.html#9); if the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html); there will be no monetary reward for this job

**Comment by 0crat** [comment]
The architect of the project has changed; @llorllale[/z](https://www.0crat.com/u/llorllale) is not at this role anymore; @paulodamaso[/z](https://www.0crat.com/u/paulodamaso) is the architect now

**Comment by victornoel** [comment]
@paulodamaso I'm blocked by this to solve other issues in cactoos, the fix is quite simple, can this be proactively assigned to somebody and a release made if possible?

**Comment by paulodamaso** [comment]
@victornoel Can you do it for us?

**Comment by victornoel** [comment]
@paulodamaso yes

**Comment by paulodamaso** [comment]
@0crat assign @victornoel

**Comment by paulodamaso** [comment]
@victornoel Thanks!

**Comment by 0crat** [comment]
> @0crat assign @victornoel  ([here](https://github.com/llorllale/cactoos-matchers/issues/132#issuecomment-534528754))

@paulodamaso There is an unrecoverable failure on my side. Please, submit it [here](https://github.com/zerocracy/farm/issues):

```
PID: 4@da69ffec-8968-43a0-8327-1386ee9c8dff, thread: pool-26-thread-2
com.zerocracy.claims.ClaimIn[198] java.lang.IllegalArgumentException: Parameter "age" not found in "Order was canceled" among: [comment, job, issue, repo, cause, _parent_sig, flow, login]
```

[0.49.12](https://github.com/zerocracy/farm): CID: [83a83d7a-f893-4ce9-9d24-3d070b7eb495](https://www.0crat.com/footprint/C63314D6Z/83a83d7a-f893-4ce9-9d24-3d070b7eb495), Type: "Order was canceled"

**Comment by 0crat** [comment]
Manual assignment of issues is discouraged, see [§19](http://www.zerocracy.com/policy.html#19): -5 point(s) just awarded to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso)

**Comment by victornoel** [comment]
@0crat wait for #147

**Comment by 0crat** [comment]
> @0crat wait for #147 ([here](https://github.com/llorllale/cactoos-matchers/issues/132#issuecomment-534677365))

@victornoel The impediment for [#132](https://github.com/llorllale/cactoos-matchers/issues/132) was registered successfully by @victornoel[/z](https://www.0crat.com/u/victornoel)

**Comment by victornoel** [comment]
@paulodamaso can we get a new release of cactoos-matchers now that this has been merged?

**Comment by 0crat** [comment]
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @victornoel[/z](https://www.0crat.com/u/victornoel), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#132](https://github.com/llorllale/cactoos-matchers/issues/132) is now out of scope

**Comment by sereshqua** [comment]
@victornoel we will need to make sure that problem reporter and problem solver are two different persons next time

**Comment by victornoel** [comment]
@sereshqua yes I agree

**Comment by sereshqua** [comment]
@0crat quality acceptable

**Comment by 0crat** [comment]
Quality review completed: +4 point(s) just awarded to @sereshqua[/z](https://www.0crat.com/u/sereshqua)

**Comment by 0crat** [comment]
Order was finished, quality is "acceptable": +30 point(s) just awarded to @victornoel[/z](https://www.0crat.com/u/victornoel)

**Comment by paulodamaso** [comment]
@

> @paulodamaso can we get a new release of cactoos-matchers now that this has been merged?

@victornoel of course

**Comment by paulodamaso** [comment]
@rultor release tag is `0.18`

**Comment by victornoel** [comment]
@paulodamaso thx
